File: js/src/tools/memory.ts
/**
 * MemoryTool — recall what was on screen at any point in the past.
 *
 * Backed by the local screen-memory index in `~/.opendesk/memory` (see
 * src/memory/).  The index is populated by the background daemon
 * (`opendesk-js memory start`); this tool reads it plus a few control
 * actions (pause / resume / deny list / delete).  Always local — screen
 * memory never leaves the machine it was recorded on.
 */

import { Tool, ToolContext, ToolResult, checkPermission, type Attachment } from "./base.js";

const MAX_TEXT_CHARS = 6000;

export interface MemoryToolOptions {
  /** Override the opendesk home (default ~/.opendesk or $OPENDESK_HOME). */
  home?: string;
}

export class MemoryTool extends Tool {
  name = "memory";
  description =
    "Recall what the user saw on screen in the past. A local background daemon " +
    "(`opendesk-js memory start`) captures the screen every ~30s, OCRs it, and " +
    "indexes the text with a thumbnail. Nothing is uploaded.\n\n" +
    "Actions:\n" +
    "  search   — full-text search. query='error', since='tuesday', app='Terminal'. " +
    "Returns matching moments (id, time, app, window title, snippet).\n" +
    "  show     — fetch one moment by id: full OCR text + thumbnail image.\n" +
    "  timeline — list captured moments in a period, optionally filtered by app " +
    "(e.g. 'every time the dashboard was open this month').\n" +
    "  status   — is the daemon running/paused, how much is stored, config.\n" +
    "  pause / resume — stop or restart capture (pause accepts duration='1h').\n" +
    "  deny     — manage the per-app deny list (denyAdd / denyRemove).\n" +
    "  config   — change intervalSeconds, storageCapMb, retentionDays.\n" +
    "  delete   — remove stored moments for a time range and/or app.\n\n" +
    "Time phrases for since/until: '2h', '3d', 'yesterday', 'tuesday', " +
    "'last week', 'this month', 'yesterday afternoon', '2026-09-01'.\n" +
    "Typical flow: search → pick an id → show to read the full text / image.";

  schema = {
    type: "object",
    required: ["action"],
    properties: {
      action: {
        type: "string",
        enum: ["search", "show", "timeline", "status", "pause", "resume", "deny", "config", "delete"],
        description: "What to do. See tool description.",
      },
      query: { type: "string", description: "Search terms (action=search). All terms must appear; quote phrases; 'invoice*' for prefix." },
      since: { type: "string", description: "Start of the time window: '2h', 'yesterday', 'tuesday', 'last week', ISO date." },
      until: { type: "string", description: "End of the time window (same formats as since)." },
      app: { type: "string", description: "Only moments whose app name or window title contains this (case-insensitive)." },
      limit: { type: "integer", minimum: 1, maximum: 200, default: 15, description: "Max results for search/timeline." },
      id: { type: "integer", description: "Frame id (action=show)." },
      includeImage: { type: "boolean", default: true, description: "For action=show: attach the thumbnail image." },
      duration: { type: "string", description: "For action=pause: how long ('30m', '2h'). Omit to pause until resumed." },
      denyAdd: { type: "string", description: "For action=deny: app name / title substring to add." },
      denyRemove: { type: "string", description: "For action=deny: entry to remove." },
      intervalSeconds: { type: "number", minimum: 5, description: "For action=config: seconds between captures." },
      storageCapMb: { type: "integer", minimum: 16, description: "For action=config: storage ceiling in MB." },
      retentionDays: { type: "integer", minimum: 1, description: "For action=config: delete frames older than this." },
      confirm: { type: "boolean", default: false, description: "For action=delete: must be true to actually delete." },
    },
  };

  private home?: string;

  constructor(opts: MemoryToolOptions = {}) {
    super();
    this.home = opts.home;
  }

  private resolveHome(ctx: ToolContext): string | undefined {
    if (this.home) return this.home;
    const meta = (ctx as { metadata?: Record<string, unknown> }).metadata;
    const h = meta?.["opendeskHome"] ?? meta?.["opendesk_home"];
    return typeof h === "string" && h ? h : undefined;
  }

  async execute(ctx: ToolContext, params: Record<string, unknown>): Promise<ToolResult> {
    const p = params as Params;
    const action = String(p.action ?? "");
    await checkPermission(ctx, "memory", action, `Screen memory: ${action}${p.query ? ` '${p.query}'` : ""}`);
    const home = this.resolveHome(ctx);

    try {
      switch (action) {
        case "search": return await this.search(home, p);
        case "show": return await this.show(home, p);
        case "timeline": return await this.timeline(home, p);
        case "status": return await this.status(home);
        case "pause": return await this.pause(home, p);
        case "resume": return await this.resume(home);
        case "deny": return await this.deny(home, p);
        case "config": return await this.configure(home, p);
        case "delete": return await this.remove(home, p);
        default: return this.err("Memory error", `Unknown action '${action}'`);
      }
    } catch (e) {
      return this.err("Memory error", e instanceof Error ? e.message : String(e));
    }
  }

  // -- search / show / timeline ------------------------------------------------

  private async search(home: string | undefined, p: Params): Promise<ToolResult> {
    const { MemoryStore } = await import("../memory/store.js");
    const { fmtRange, parseRange } = await import("../memory/timeparse.js");
    if (!p.query || !String(p.query).trim()) {
      throw new Error("query is required for action='search' (use action='timeline' to browse).");
    }
    const [start, end] = parseRange(p.since, p.until);
    const store = new MemoryStore(home);
    const frames = store.search(String(p.query), { start, end, app: p.app ?? null, limit: limitOf(p) });
    const hint = await this.emptyHint(store, home);
    let header = `Screen memory search '${p.query}' — ${fmtRange(start, end)}`;
    if (p.app) header += ` — app~'${p.app}'`;
    if (!frames.length) {
      return this.ok("Memory: no matches", `${header}\nNo matching moments.${hint}`, [], { count: 0 });
    }
    const lines = [header, `${frames.length} match(es), best first:`, ""];
    for (const f of frames) lines.push(frameLine(f, true));
    lines.push("", "Use memory(action='show', id=<id>) for the full text and thumbnail.");
    return this.ok(`Memory: ${frames.length} match(es)`, lines.join("\n"), [], { count: frames.length, ids: frames.map((f) => f.id) });
  }

  private async timeline(home: string | undefined, p: Params): Promise<ToolResult> {
    const { MemoryStore } = await import("../memory/store.js");
    const { fmtRange, parseRange } = await import("../memory/timeparse.js");
    const [start, end] = parseRange(p.since, p.until);
    const store = new MemoryStore(home);
    const frames = store.timeline({ start, end, app: p.app ?? null, limit: limitOf(p) });
    const apps = store.apps({ start, end });
    const hint = await this.emptyHint(store, home);
    let header = `Screen memory timeline — ${fmtRange(start, end)}`;
    if (p.app) header += ` — app~'${p.app}'`;
    if (!frames.length) {
      return this.ok("Memory: nothing recorded", `${header}\nNo moments recorded in this window.${hint}`, [], { count: 0 });
    }
    const lines = [header, ""];
    if (apps.length && !p.app) {
      lines.push("Apps in window: " + apps.slice(0, 8).map(([a, n]) => `${a || "(unknown)"} ×${n}`).join(", "), "");
    }
    lines.push(`Most recent ${frames.length}:`);
    for (const f of frames) lines.push(frameLine(f, false));
    lines.push("", "Use memory(action='show', id=<id>) to read a moment in full.");
    return this.ok(`Memory timeline (${frames.length})`, lines.join("\n"), [], { count: frames.length, ids: frames.map((f) => f.id) });
  }

  private async show(home: string | undefined, p: Params): Promise<ToolResult> {
    const { MemoryStore, frameWhen } = await import("../memory/store.js");
    if (p.id === undefined || p.id === null) throw new Error("id is required for action='show'.");
    const store = new MemoryStore(home);
    const frame = store.get(Number(p.id));
    if (!frame) return this.err("Memory: not found", `No moment with id ${p.id}.`);
    const includeImage = p.includeImage !== false;
    const thumb = includeImage ? store.thumbnail(frame) : null;

    let text = frame.text.trim() || "(no text was recognised in this frame)";
    let truncated = "";
    if (text.length > MAX_TEXT_CHARS) {
      text = text.slice(0, MAX_TEXT_CHARS);
      truncated = `\n… [truncated, ${frame.text.length} chars total]`;
    }
    const lines = [`Moment #${frame.id} — ${frameWhen(frame)}`, `App:    ${frame.app || "(unknown)"}`];
    if (frame.title) lines.push(`Window: ${frame.title}`);
    lines.push("", "Screen text:", text + truncated);
    const attachments: Attachment[] = thumb
      ? [{ filename: `memory-${frame.id}.jpg`, mediaType: "image/jpeg", content: thumb }]
      : [];
    return this.ok(`Memory #${frame.id} (${frame.app || "unknown"})`, lines.join("\n"), attachments, {
      id: frame.id, ts: frame.ts, when: frameWhen(frame), app: frame.app, title: frame.title, text: frame.text, thumb_path: frame.thumb_path,
    });
  }

  // -- status / control --------------------------------------------------------

  private async status(home: string | undefined): Promise<ToolResult> {
    const { MemoryStore } = await import("../memory/store.js");
    const { daemonAlive, describePause, getPause, loadConfig, readDaemonState } = await import("../memory/config.js");
    const { fmtTs } = await import("../memory/timeparse.js");
    const cfg = loadConfig(home);
    const pause = getPause(home);
    const alive = daemonAlive(home);
    const state: Partial<import("../memory/config.js").DaemonState> = readDaemonState(home) ?? {};
    const store = new MemoryStore(home);
    const st = store.stats();
    const daemonLine = alive
      ? `running (pid ${state.pid}, last tick: ${state.status ?? "?"})`
      : "NOT running — start it with `opendesk-js memory start`";
    const span = st.frames && st.oldest !== null && st.newest !== null ? `  (${fmtTs(st.oldest)} → ${fmtTs(st.newest)})` : "";
    const lines = [
      "Screen memory status",
      `  daemon:    ${daemonLine}`,
      `  capture:   ${pause ? "PAUSED — " + describePause(pause) : "active"}`,
      `  store:     ${store.dir}`,
      `  frames:    ${st.frames}${span}`,
      `  size:      ${mb(st.totalBytes)} MB of ${cfg.storage_cap_mb} MB cap (thumbnails ${mb(st.thumbBytes)} MB, index ${mb(st.indexBytes)} MB)`,
      `  retention: ${cfg.retention_days} days`,
      `  interval:  every ${cfg.interval_seconds}s`,
      `  deny list: ${cfg.deny_apps.length ? cfg.deny_apps.join(", ") : "(empty)"}`,
      `  hotkey:    ${cfg.pause_hotkey || "(disabled)"}`,
    ];
    if (st.apps.length) lines.push("  top apps:  " + st.apps.slice(0, 6).map(([a, n]) => `${a || "(unknown)"} ×${n}`).join(", "));
    return this.ok("Memory status", lines.join("\n"), [], {
      daemonAlive: alive, paused: pause !== null, frames: st.frames, bytes: st.totalBytes, config: cfg.toDict(),
    });
  }

  private async pause(home: string | undefined, p: Params): Promise<ToolResult> {
    const { describePause, parseDuration, setPause } = await import("../memory/config.js");
    const secs = p.duration ? parseDuration(String(p.duration)) : undefined;
    const state = setPause(home, { durationSeconds: secs, reason: "tool" });
    return this.ok("Memory paused", `Screen memory capture ${describePause(state)}.`);
  }

  private async resume(home: string | undefined): Promise<ToolResult> {
    const { clearPause } = await import("../memory/config.js");
    const cleared = clearPause(home);
    return this.ok("Memory resumed", cleared ? "Screen memory capture resumed." : "Capture was not paused.");
  }

  private async deny(home: string | undefined, p: Params): Promise<ToolResult> {
    const { loadConfig, saveConfig } = await import("../memory/config.js");
    const cfg = loadConfig(home);
    const msgs: string[] = [];
    if (p.denyAdd) msgs.push(cfg.denyAdd(String(p.denyAdd)) ? `Added '${p.denyAdd}'.` : `'${p.denyAdd}' already listed.`);
    if (p.denyRemove) msgs.push(cfg.denyRemove(String(p.denyRemove)) ? `Removed '${p.denyRemove}'.` : `'${p.denyRemove}' not found.`);
    if (p.denyAdd || p.denyRemove) saveConfig(cfg, home);
    const listing = cfg.deny_apps.map((d) => `  - ${d}`).join("\n") || "  (empty)";
    const out = (msgs.length ? msgs.join("\n") + "\n\n" : "") + "Deny list (apps / window titles never captured):\n" + listing;
    return this.ok("Memory deny list", out, [], { denyApps: cfg.deny_apps });
  }

  private async configure(home: string | undefined, p: Params): Promise<ToolResult> {
    const { loadConfig, saveConfig } = await import("../memory/config.js");
    const cfg = loadConfig(home);
    const changed: string[] = [];
    if (p.intervalSeconds !== undefined) { cfg.interval_seconds = Number(p.intervalSeconds); changed.push(`interval_seconds=${cfg.interval_seconds}`); }
    if (p.storageCapMb !== undefined) { cfg.storage_cap_mb = Math.floor(Number(p.storageCapMb)); changed.push(`storage_cap_mb=${cfg.storage_cap_mb}`); }
    if (p.retentionDays !== undefined) { cfg.retention_days = Math.floor(Number(p.retentionDays)); changed.push(`retention_days=${cfg.retention_days}`); }
    if (changed.length) saveConfig(cfg, home);
    const lines: string[] = [];
    if (changed.length) lines.push("Updated: " + changed.join(", "), "(the running daemon picks this up within a few ticks)", "");
    lines.push("Current config:");
    for (const [k, v] of Object.entries(cfg.toDict())) lines.push(`  ${k}: ${Array.isArray(v) ? JSON.stringify(v) : v}`);
    return this.ok("Memory config", lines.join("\n"), [], cfg.toDict() as unknown as Record<string, unknown>);
  }

  private async remove(home: string | undefined, p: Params): Promise<ToolResult> {
    const { MemoryStore } = await import("../memory/store.js");
    const { fmtRange, parseRange } = await import("../memory/timeparse.js");
    if (!(p.since || p.until || p.app)) {
      throw new Error("Refusing to delete everything: pass since/until and/or app. Use the CLI `opendesk-js memory clear` to wipe the whole index.");
    }
    const [start, end] = parseRange(p.since, p.until);
    const scope = fmtRange(start, end) + (p.app ? `, app~'${p.app}'` : "");
    const store = new MemoryStore(home);
    const candidates = store.timeline({ start, end, app: p.app ?? null, limit: Infinity });
    if (!p.confirm) {
      return this.ok("Memory delete (dry run)", `${candidates.length} moment(s) would be deleted for ${scope}. Re-run with confirm=true to delete.`, [], { count: candidates.length });
    }
    const n = store.delete(candidates.map((f) => f.id));
    return this.ok("Memory deleted", `Deleted ${n} moment(s) for ${scope}.`, [], { count: n });
  }

  private async emptyHint(store: { stats(): { frames: number } }, home?: string): Promise<string> {
    const { daemonAlive, getPause } = await import("../memory/config.js");
    if (store.stats().frames === 0) {
      if (!daemonAlive(home)) return "\nThe index is empty and the capture daemon is not running. Start it with `opendesk-js memory start`.";
      return "\nThe index is empty — the daemon has only just started.";
    }
    if (getPause(home) !== null) return "\n(Capture is currently paused.)";
    return "";
  }
}

// ---------------------------------------------------------------------------

interface Params {
  action?: string;
  query?: string;
  since?: string;
  until?: string;
  app?: string;
  limit?: number;
  id?: number;
  includeImage?: boolean;
  duration?: string;
  denyAdd?: string;
  denyRemove?: string;
  intervalSeconds?: number;
  storageCapMb?: number;
  retentionDays?: number;
  confirm?: boolean;
}

function limitOf(p: Params): number {
  const n = Number(p.limit);
  if (!Number.isFinite(n) || n < 1) return 15;
  return Math.min(200, Math.floor(n));
}

function frameLine(f: import("../memory/store.js").Frame, snippet: boolean): string {
  const app = f.app || "(unknown)";
  const title = f.title && f.title !== f.app ? ` — ${f.title.slice(0, 70)}` : "";
  const when = new Date(f.ts * 1000);
  const pad = (n: number) => String(n).padStart(2, "0");
  const stamp = `${when.getFullYear()}-${pad(when.getMonth() + 1)}-${pad(when.getDate())} ${pad(when.getHours())}:${pad(when.getMinutes())}:${pad(when.getSeconds())}`;
  let line = `  #${String(f.id).padEnd(6)} ${stamp}  [${app}]${title}`;
  if (snippet) {
    const snip = (f.snippet || f.text.slice(0, 120)).replace(/\n/g, " ").trim();
    if (snip) line += `\n          ${snip.slice(0, 220)}`;
  }
  return line;
}

function mb(n: number): string {
  return (n / (1024 * 1024)).toFixed(1);
}

File: python/opendesk/tools/memory.py
"""MemoryTool — recall what was on screen at any point in the past.

Backed by the local screen-memory index in ``~/.opendesk/memory`` (see
:mod:`opendesk.memory`).  The index is populated by the background daemon
(``opendesk memory start``); this tool only *reads* it, plus a few control
actions (pause / resume / deny list / delete).

The tool always talks to the *local* index — screen memory never leaves
the machine it was recorded on, so it is intentionally not peer-routable.
"""

from __future__ import annotations

import datetime as _dt
from pathlib import Path
from typing import Any, List, Literal, Optional

from pydantic import Field

from opendesk.tools.base import Attachment, Tool, ToolContext, ToolResult

_MAX_TEXT_CHARS = 6000


class MemoryTool(Tool):
    """Search and manage the local screen-memory history."""

    name = "memory"
    description = (
        "Recall what the user saw on screen in the past. A local background "
        "daemon (`opendesk memory start`) captures the screen every ~30s, OCRs it, "
        "and indexes the text with a thumbnail. Nothing is uploaded.\n\n"
        "Actions:\n"
        "  search   — full-text search. query='error', since='tuesday', app='Terminal'. "
        "Returns matching moments (id, time, app, window title, snippet).\n"
        "  show     — fetch one moment by id: full OCR text + thumbnail image.\n"
        "  timeline — list captured moments in a period, optionally filtered by app "
        "(e.g. 'every time the dashboard was open this month').\n"
        "  status   — is the daemon running/paused, how much is stored, config.\n"
        "  pause / resume — stop or restart capture (pause accepts duration='1h').\n"
        "  deny     — manage the per-app deny list (deny_add / deny_remove).\n"
        "  config   — change interval_seconds, storage_cap_mb, retention_days.\n"
        "  delete   — remove stored moments for a time range and/or app.\n\n"
        "Time phrases for since/until: '2h', '3d', 'yesterday', 'tuesday', "
        "'last week', 'this month', 'yesterday afternoon', '2026-09-01'.\n"
        "Typical flow: search → pick an id → show to read the full text / image."
    )

    class Params(Tool.Params):
        action: Literal[
            "search", "show", "timeline", "status", "pause", "resume",
            "deny", "config", "delete",
        ] = Field(description="What to do. See tool description.")
        query: Optional[str] = Field(
            default=None,
            description=(
                "Search terms (action=search). All terms must appear; quote phrases. "
                "Append * for prefix match (e.g. 'invoice*')."
            ),
        )
        since: Optional[str] = Field(
            default=None,
            description="Start of the time window: '2h', 'yesterday', 'tuesday', 'last week', ISO date.",
        )
        until: Optional[str] = Field(
            default=None,
            description="End of the time window (same formats as since).",
        )
        app: Optional[str] = Field(
            default=None,
            description="Only moments whose app name or window title contains this (case-insensitive).",
        )
        limit: int = Field(default=15, ge=1, le=200, description="Max results for search/timeline.")
        id: Optional[int] = Field(default=None, description="Frame id (action=show).")
        include_image: bool = Field(
            default=True, description="For action=show: attach the thumbnail image.",
        )
        duration: Optional[str] = Field(
            default=None,
            description="For action=pause: how long ('30m', '2h'). Omit to pause until resumed.",
        )
        deny_add: Optional[str] = Field(
            default=None, description="For action=deny: app name / title substring to add.",
        )
        deny_remove: Optional[str] = Field(
            default=None, description="For action=deny: entry to remove.",
        )
        interval_seconds: Optional[float] = Field(
            default=None, ge=5, description="For action=config: seconds between captures.",
        )
        storage_cap_mb: Optional[int] = Field(
            default=None, ge=16, description="For action=config: storage ceiling in MB.",
        )
        retention_days: Optional[int] = Field(
            default=None, ge=1, description="For action=config: delete frames older than this.",
        )
        confirm: bool = Field(
            default=False,
            description="For action=delete: must be true to actually delete.",
        )

    def __init__(self, home: Optional[Path] = None) -> None:
        self._home = home

    # ------------------------------------------------------------------

    def _resolve_home(self, ctx: ToolContext) -> Optional[Path]:
        if self._home is not None:
            return self._home
        meta_home = ctx.metadata.get("opendesk_home") if ctx.metadata else None
        if meta_home:
            return Path(str(meta_home)).expanduser()
        import os
        env = os.environ.get("OPENDESK_HOME")
        return Path(env).expanduser() if env else None

    async def execute(self, ctx: ToolContext, params: "MemoryTool.Params") -> ToolResult:
        action = params.action
        await ctx.check_permission(
            tool="memory", argument=action,
            description=f"Screen memory: {action}"
            + (f" {params.query!r}" if params.query else ""),
        )
        home = self._resolve_home(ctx)

        try:
            if action == "search":
                return self._search(home, params)
            if action == "show":
                return self._show(home, params)
            if action == "timeline":
                return self._timeline(home, params)
            if action == "status":
                return self._status(home)
            if action == "pause":
                return self._pause(home, params)
            if action == "resume":
                return self._resume(home)
            if action == "deny":
                return self._deny(home, params)
            if action == "config":
                return self._config(home, params)
            if action == "delete":
                return self._delete(home, params)
        except ValueError as exc:
            return ToolResult(title="Memory error", output=str(exc), error=True)
        except ImportError as exc:
            return ToolResult(title="Memory error", output=str(exc), error=True)
        return ToolResult(title="Memory error", output=f"Unknown action {action!r}", error=True)

    # -- search / show / timeline ---------------------------------------------

    def _search(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.store import MemoryStore
        from opendesk.memory.timeparse import fmt_range, parse_range

        if not p.query or not p.query.strip():
            raise ValueError("query is required for action='search' (use action='timeline' to browse).")
        start, end = parse_range(p.since, p.until)
        with MemoryStore(home) as store:
            frames = store.search(p.query, start=start, end=end, app=p.app, limit=p.limit)
            empty_hint = self._empty_hint(store, home)

        header = f"Screen memory search {p.query!r} — {fmt_range(start, end)}"
        if p.app:
            header += f" — app~{p.app!r}"
        if not frames:
            return ToolResult(
                title="Memory: no matches",
                output=f"{header}\nNo matching moments.{empty_hint}",
                metadata={"count": 0},
            )
        lines = [header, f"{len(frames)} match(es), best first:", ""]
        for f in frames:
            lines.append(_frame_line(f))
        lines.append("")
        lines.append("Use memory(action='show', id=<id>) for the full text and thumbnail.")
        return ToolResult(
            title=f"Memory: {len(frames)} match(es)",
            output="\n".join(lines),
            metadata={"count": len(frames), "ids": [f.id for f in frames]},
        )

    def _timeline(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.store import MemoryStore
        from opendesk.memory.timeparse import fmt_range, parse_range

        start, end = parse_range(p.since, p.until)
        with MemoryStore(home) as store:
            frames = store.timeline(start=start, end=end, app=p.app, limit=p.limit)
            apps = store.apps(start=start, end=end)
            empty_hint = self._empty_hint(store, home)

        header = f"Screen memory timeline — {fmt_range(start, end)}"
        if p.app:
            header += f" — app~{p.app!r}"
        if not frames:
            return ToolResult(
                title="Memory: nothing recorded",
                output=f"{header}\nNo moments recorded in this window.{empty_hint}",
                metadata={"count": 0},
            )
        lines = [header, ""]
        if apps and not p.app:
            top = ", ".join(f"{a or '(unknown)'} ×{n}" for a, n in apps[:8])
            lines.append(f"Apps in window: {top}")
            lines.append("")
        lines.append(f"Most recent {len(frames)}:")
        for f in frames:
            lines.append(_frame_line(f, snippet=False))
        lines.append("")
        lines.append("Use memory(action='show', id=<id>) to read a moment in full.")
        return ToolResult(
            title=f"Memory timeline ({len(frames)})",
            output="\n".join(lines),
            metadata={"count": len(frames), "ids": [f.id for f in frames]},
        )

    def _show(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.store import MemoryStore

        if p.id is None:
            raise ValueError("id is required for action='show'.")
        with MemoryStore(home) as store:
            frame = store.get(int(p.id))
            thumb = store.thumbnail(frame) if (frame and p.include_image) else None
        if frame is None:
            return ToolResult(title="Memory: not found", output=f"No moment with id {p.id}.", error=True)

        text = frame.text.strip() or "(no text was recognised in this frame)"
        truncated = ""
        if len(text) > _MAX_TEXT_CHARS:
            text = text[:_MAX_TEXT_CHARS]
            truncated = f"\n… [truncated, {len(frame.text)} chars total]"
        lines = [
            f"Moment #{frame.id} — {frame.when}",
            f"App:    {frame.app or '(unknown)'}",
        ]
        if frame.title:
            lines.append(f"Window: {frame.title}")
        lines.append("")
        lines.append("Screen text:")
        lines.append(text + truncated)
        attachments = []
        if thumb:
            attachments.append(Attachment(f"memory-{frame.id}.jpg", thumb, "image/jpeg"))
        return ToolResult(
            title=f"Memory #{frame.id} ({frame.app or 'unknown'})",
            output="\n".join(lines),
            attachments=attachments,
            metadata=frame.to_dict(),
        )

    # -- status / control ----------------------------------------------------

    def _status(self, home: Optional[Path]) -> ToolResult:
        from opendesk.memory.config import daemon_alive, get_pause, load_config, read_daemon_state
        from opendesk.memory.store import MemoryStore
        from opendesk.memory.timeparse import fmt_ts

        cfg = load_config(home)
        pause = get_pause(home)
        alive = daemon_alive(home)
        state = read_daemon_state(home) or {}
        with MemoryStore(home) as store:
            st = store.stats()
            fts = store.has_fts
            store_dir = store.dir

        if alive:
            daemon_line = f"running (pid {state.get('pid')}, last tick: {state.get('status', '?')})"
        else:
            daemon_line = "NOT running — start it with `opendesk memory start`"
        lines = [
            "Screen memory status",
            f"  daemon:    {daemon_line}",
            f"  capture:   {'PAUSED — ' + pause.describe() if pause else 'active'}",
            f"  store:     {store_dir}",
            f"  frames:    {st.frames}"
            + (f"  ({fmt_ts(st.oldest)} → {fmt_ts(st.newest)})" if st.frames and st.oldest and st.newest else ""),
            f"  size:      {_mb(st.total_bytes)} MB of {cfg.storage_cap_mb} MB cap "
            f"(thumbnails {_mb(st.thumb_bytes)} MB, index {_mb(st.db_bytes)} MB)",
            f"  retention: {cfg.retention_days} days",
            f"  interval:  every {cfg.interval_seconds:g}s",
            f"  deny list: {', '.join(cfg.deny_apps) if cfg.deny_apps else '(empty)'}",
            f"  hotkey:    {cfg.pause_hotkey or '(disabled)'}",
            f"  search:    {'FTS5 full-text' if fts else 'LIKE fallback (no FTS5)'}",
        ]
        if st.apps:
            lines.append("  top apps:  " + ", ".join(f"{a or '(unknown)'} ×{n}" for a, n in st.apps[:6]))
        return ToolResult(
            title="Memory status",
            output="\n".join(lines),
            metadata={
                "daemon_alive": alive, "paused": pause is not None, "frames": st.frames,
                "bytes": st.total_bytes, "config": cfg.to_dict(),
            },
        )

    def _pause(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.config import parse_duration, set_pause

        secs = parse_duration(p.duration) if p.duration else None
        state = set_pause(home, duration_seconds=secs, reason="tool")
        return ToolResult(title="Memory paused", output=f"Screen memory capture {state.describe()}.")

    def _resume(self, home: Optional[Path]) -> ToolResult:
        from opendesk.memory.config import clear_pause

        cleared = clear_pause(home)
        return ToolResult(
            title="Memory resumed",
            output="Screen memory capture resumed." if cleared else "Capture was not paused.",
        )

    def _deny(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.config import load_config, save_config

        cfg = load_config(home)
        msgs: list[str] = []
        if p.deny_add:
            msgs.append(
                f"Added {p.deny_add!r}." if cfg.deny_add(p.deny_add) else f"{p.deny_add!r} already listed."
            )
        if p.deny_remove:
            msgs.append(
                f"Removed {p.deny_remove!r}." if cfg.deny_remove(p.deny_remove) else f"{p.deny_remove!r} not found."
            )
        if p.deny_add or p.deny_remove:
            save_config(cfg, home)
        listing = "\n".join(f"  - {d}" for d in cfg.deny_apps) or "  (empty)"
        out = ("\n".join(msgs) + "\n\n" if msgs else "") + (
            "Deny list (apps / window titles never captured):\n" + listing
        )
        return ToolResult(title="Memory deny list", output=out, metadata={"deny_apps": cfg.deny_apps})

    def _config(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.config import load_config, save_config

        cfg = load_config(home)
        changed: list[str] = []
        if p.interval_seconds is not None:
            cfg.interval_seconds = float(p.interval_seconds)
            changed.append(f"interval_seconds={cfg.interval_seconds:g}")
        if p.storage_cap_mb is not None:
            cfg.storage_cap_mb = int(p.storage_cap_mb)
            changed.append(f"storage_cap_mb={cfg.storage_cap_mb}")
        if p.retention_days is not None:
            cfg.retention_days = int(p.retention_days)
            changed.append(f"retention_days={cfg.retention_days}")
        if changed:
            save_config(cfg, home)
        lines = []
        if changed:
            lines.append("Updated: " + ", ".join(changed))
            lines.append("(the running daemon picks this up within a few ticks)")
            lines.append("")
        lines.append("Current config:")
        for k, v in cfg.to_dict().items():
            lines.append(f"  {k}: {v}")
        return ToolResult(title="Memory config", output="\n".join(lines), metadata=cfg.to_dict())

    def _delete(self, home: Optional[Path], p: "MemoryTool.Params") -> ToolResult:
        from opendesk.memory.store import MemoryStore
        from opendesk.memory.timeparse import fmt_range, parse_range

        if not (p.since or p.until or p.app):
            raise ValueError(
                "Refusing to delete everything: pass since/until and/or app. "
                "Use the CLI `opendesk memory clear` to wipe the whole index."
            )
        start, end = parse_range(p.since, p.until)
        scope = fmt_range(start, end) + (f", app~{p.app!r}" if p.app else "")
        with MemoryStore(home) as store:
            candidates = store.timeline(start=start, end=end, app=p.app, limit=100000)
            if not p.confirm:
                return ToolResult(
                    title="Memory delete (dry run)",
                    output=(
                        f"{len(candidates)} moment(s) would be deleted for {scope}. "
                        "Re-run with confirm=true to delete."
                    ),
                    metadata={"count": len(candidates)},
                )
            n = store.delete([f.id for f in candidates])
        return ToolResult(
            title="Memory deleted",
            output=f"Deleted {n} moment(s) for {scope}.",
            metadata={"count": n},
        )

    # ------------------------------------------------------------------

    @staticmethod
    def _empty_hint(store: Any, home: Optional[Path]) -> str:
        from opendesk.memory.config import daemon_alive, get_pause

        if store.stats().frames == 0:
            if not daemon_alive(home):
                return (
                    "\nThe index is empty and the capture daemon is not running. "
                    "Start it with `opendesk memory start`."
                )
            return "\nThe index is empty — the daemon has only just started."
        if get_pause(home) is not None:
            return "\n(Capture is currently paused.)"
        return ""


# ---------------------------------------------------------------------------


def _frame_line(f: Any, *, snippet: bool = True) -> str:
    app = f.app or "(unknown)"
    title = f" — {f.title[:70]}" if f.title and f.title != f.app else ""
    line = f"  #{f.id:<6} {f.when}  [{app}]{title}"
    if snippet:
        snip = (f.snippet or f.text[:120]).replace("\n", " ").strip()
        if snip:
            line += f"\n          {snip[:220]}"
    return line


def _mb(n: int) -> str:
    return f"{n / (1024 * 1024):.1f}"
